### PR TITLE
drivers: counter: add Maxim DS3231 support

### PR DESCRIPTION
**Pull request by pabigot**
This PR provides support for the [Maxim DS3231 TCXO RTC](https://www.maximintegrated.com/en/products/analog/real-time-clocks/DS3231.html), a high-precision real-time clock peripheral with support for battery backup.

This is implemented as a counter, but the DS3231 has significant functionality enhancements beyond a general RTC API.    Also, as an I2C peripheral it has limitations that are incompatible with assumptions about how a counter can function.  Most users of its advanced functionality will interact with it through device specific API provided in a driver-specific header following the approach summarized in #17072.  Consequently the interface file is placed in the (existing) `include/drivers/rtc` directory, and the extension functions are named based on it being a specialization of a (no longer existing) `rtc` driver class.

There is a question of where to collect the documentation.  Because it's device specific, it really should be its own page at API Reference : Peripherals : RTC : DS3231.  For now I've just put it under (newly created) RTC.

As more infrastructure is added to support management of civil time tracking in Zephyr some of the functionality in this driver may need to be generalized. For example, the `rtc_ds3231_syncpoint` structure should probably be generalized and placed in the recently created `<sys/timeutil.h>` header long with functions that support generic conversion from a synchronization point and local time to RTC and back.

### CODE REVIEW

File: doc/reference/peripherals/rtc.rst
Comment (dbkinder) [inline doc/reference/peripherals/rtc.rst]: Missing an overview here...

File: samples/drivers/rtc_ds3231/README.rst
Comment (dbkinder) [inline samples/drivers/rtc_ds3231/README.rst]: use two commas (``,``) in this sentence instead of two semicolons (``;``)

File: samples/drivers/rtc_ds3231/README.rst
Comment (dbkinder) [inline samples/drivers/rtc_ds3231/README.rst]: If you write this like so, it will link to the config option documentation:
```
Next, if the sample application option :option:`CONFIG_APP_SET_ALIGNED_CLOCK`
is set, the civil time will be advanced to ...
```

File: samples/drivers/rtc_ds3231/README.rst
Comment (dbkinder) [inline samples/drivers/rtc_ds3231/README.rst]: into ``the`` run

File: samples/drivers/rtc_ds3231/README.rst
Comment (dbkinder) [inline samples/drivers/rtc_ds3231/README.rst]: ``the`` corresponding

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: doc/reference/peripherals/rtc.rst
Comment (pabigot) [reply doc/reference/peripherals/rtc.rst]: Yeah, I just copied counter.  The only peripherals that have overview are I2C and sensors.

Part of the reason I can't add one is that this is for RTC peripherals in general, and the PR provides a device-specific API.  I need instruction on how to document that.  From the PR description:
> A final question is where to collect the documentation. Because it's device specific, it really should be its own page at API Reference : Peripherals : RTC : DS3231. For now I've just put it under (newly created) RTC.

I've applied the other changes; thanks.

**Review by dbkinder — APPROVED** [review]
Sigh.  OK for now.  I'll add checking for overviews for the API docs to the bin list :)

**Review by dbkinder — CHANGES_REQUESTED** [review]
The doc build is failing because there's no visible option for CONFIG_APP_SET_ALIGNED_CLOCK defined in the Kconfig tree?

File: samples/drivers/rtc_ds3231/README.rst
Comment (dbkinder) [inline samples/drivers/rtc_ds3231/README.rst]: Sorry, my suggestion to use :option:`CONFIG_APP_ALIGNED_CLOCK` doesn't work when the config option is a "private" config option for a sample (and not part of the big KConfig tree).  So this would need to be back to the way you had it as simply 
```
Next, if the sample application option ``CONFIG_APP_SET_ALIGNED_CLOCK`` is set ...
```

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: samples/drivers/rtc_ds3231/README.rst
Comment (pabigot) [reply samples/drivers/rtc_ds3231/README.rst]: Reverted.

**Review by dbkinder — APPROVED** [review]
+1 for doc changes, thanks.

**Comment by carlescufi** [comment]
@anangl could you please take a look at the proposed API extension mechanism?

**Comment by pabigot** [comment]
Added to dev-review in hopes of getting feedback on this specifically, and device-specific APIs in general.  TSC mention and posting on devel didn't produce much/anything.  Base question: can this make it into 2.0.0?

File: drivers/counter/ds3231.c
Comment (nordic-krch) [inline drivers/counter/ds3231.c]: this assignment is redundant in case of absolute. What about moving it to else clause (if not absolute)?

File: drivers/counter/ds3231.c
Comment (nordic-krch) [inline drivers/counter/ds3231.c]: is it possible that alarm is canceled and callback is still triggered? in that case `handler` may be null,

**Comment by galak** [comment]
DEV-REVIEW: Marking #17072 as 'TSC' with regards to the approach for how to have driver specific api's.  This is just an implementation of those ideas.

File: drivers/counter/ds3231.c
Comment (pabigot) [reply drivers/counter/ds3231.c]: done

File: drivers/counter/ds3231.c
Comment (pabigot) [reply drivers/counter/ds3231.c]: checked at call site; thanks.

**Comment by pabigot** [comment]
Excerpted [this comment on #17072](https://github.com/zephyrproject-rtos/zephyr/pull/17072#discussion_r311144317) by @MaureenHelm:
> Implementing the generic counter interface will make people think it's a counter. If you're worried about people thinking it's a counter, then why implement the generic counter interface at all?
>
> Similarly, if you think applications will use the device-specific functions exclusively, then isn't implementing the generic counter interface a waste of code space? I think we either need to revisit the RTC driver API or you should have a driver-specific API.

The description in #17072 was based on early work.  I believe in the final version the usefulness as a counter is more clear: in particular that you can set alarms based on a `time_t` value without having to mess with the driver-specific API, and you can read the current time using the standard API.

More generally I think an RTC API (which may or may not extend Counter) may be useful, but we need more examples of capability before we can figure out what it should look like.

**Comment by pabigot** [comment]
All current review comments addressed.  Because I expect subsequent work to require extensions to civic time support in Zephyr I'm not going to push this for 2.0.0, as that would complicate the API changes.

@dbkinder Please check new commit documenting generic approach to API extension.

File: doc/reference/drivers/index.rst
Comment (dbkinder) [inline doc/reference/drivers/index.rst]: change ``defining`` to ``definition``

File: samples/drivers/rtc_ds3231/README.rst
Comment (dbkinder) [inline samples/drivers/rtc_ds3231/README.rst]: Change to ``Next, ``

**Review by dbkinder — CHANGES_REQUESTED** [review]
A couple of typos, otherwise docs look good.

**Comment by pabigot** [comment]
> A couple of typos, otherwise docs look good.

@dbkinder requested changes made.  Thanks.

File: drivers/counter/ds3231.c
Comment (pizi-nordic) [inline drivers/counter/ds3231.c]: This looks like some debug code left behind.

**Review by dbkinder — APPROVED** [review]
+1 for docs, thanks

File: drivers/counter/ds3231.c
Comment (pabigot) [reply drivers/counter/ds3231.c]: removed

**Comment by jhedberg** [comment]
Any update on this PR? Could the assigned reviewers (especially the codeowner ones) take a look please?

**Comment by jhedberg** [comment]
Is this PR waiting for something (other than more reviews) before it's ready for merging? @pabigot looks like you'll need to fix a CODEOWNERS conflict either way.

**Comment by pabigot** [comment]
This is only waiting for more reviews, but it's been waiting so long it does need to be rebased and retested.  There's more interest now in supporting civil time in Zephyr, so I'll try to do that early in the 2.3 development cycle.

**Comment by pabigot** [comment]
This has been updated again.  Please take a look; I'd like to get it into 2.3.

File: samples/drivers/rtc_ds3231/prj.conf
Comment (nordic-krch) [inline samples/drivers/rtc_ds3231/prj.conf]: it's already selected by `COUNTER_DS3231`

File: drivers/counter/ds3231.c
Comment (nordic-krch) [reply drivers/counter/ds3231.c]: is it?

File: include/drivers/rtc/ds3231.h
Comment (nordic-krch) [inline include/drivers/rtc/ds3231.h]: would be good to add build time assert which ensures that `counter_api` is the first item in the api struct.

**Review by nordic-krch — COMMENTED** [review]
it didn't go very deep and have no way of testings. Left few some comments.

File: include/drivers/rtc/ds3231.h
Comment (pabigot) [reply include/drivers/rtc/ds3231.h]: We don't generally do that in similar cases, like gpio which has required common prefixes.  It'd be pretty obvious if it was wrong.

File: drivers/counter/ds3231.c
Comment (pabigot) [reply drivers/counter/ds3231.c]: Was probably lost during the restoration when I deleted the branch and github wouldn't allow the PR to be reopened.  Gone now.

**Comment by GeorgeGkinis** [comment]
Nice work!

I see this PR has been waiting for 8 months.

Any chance it will be available for 2.3?

**Comment by pabigot** [comment]
> Any chance it will be available for 2.3?

Probably not, but we'll try.  The latest blocker is support for userspace, which I understand to be a requirement for all new API.  So this is going back into DNM mode until that's resolved; some background is in #24539.

**Comment by pabigot** [comment]
This has been rebased and updated with proper locking and usermode support for the extended sync functions and the underlying counter API.  I would like to see something like the `rtc_ds3231_{req,get}_syncpoint()` API in whatever we use for RTC extensions.

Please provide feedback, particularly @jdascenzio @henrikbrixandersen  @sslupsky  who have expressed interest in RTC.  I would like to get this in to 2.3, rather than attempt to beat @nordic-krch 's record of 367 days from PR open to merge.

File: drivers/counter/ds3231.c
Comment (jfischer-phytec-iot) [inline drivers/counter/ds3231.c:130]: That is unusual.

**Comment by carlescufi** [comment]
@henrikbrixandersen and @jdascenzio could you please review, especially in the context of the work for #23526

File: drivers/counter/ds3231.c
Comment (pabigot) [reply drivers/counter/ds3231.c:130]: You mean mixed case?  I suppose, though it's standard for some roles (e.g. CMSIS-style bitfield values), and is also present internally Zephyr's time_utils.h where the units suffix is (correctly) in lower case.

If it really bothers people I'll uglify it.

**Comment by pabigot** [comment]
Rebased and all public symbols renamed from `rtc_ds3231_` to `maxim_ds3231_` in anticipation of possible dedicated RTC API.

File: drivers/counter/ds3231.c
Comment (jfischer-phytec-iot) [reply drivers/counter/ds3231.c]: I meant to define a macro in middle of a struct.

File: include/drivers/rtc/maxim_ds3231.h
Comment (jfischer-phytec-iot) [inline include/drivers/rtc/maxim_ds3231.h:231]: `maxim_ds3231_alarm_callback_handler_t` is to long, please use shorter names, here `ds3231_alarm_cb_handler` is fine.

Overall just ds3231/DS3231 instead of maxim_ds3231/MAXIM_DS3231 would be sufficient.

Why not generally abbreviate `callback` with `cb` if the prefix give the only feasible interpretation like `ds3231_notify_cb`?

Comment (pabigot) [reply include/drivers/rtc/maxim_ds3231.h:231]: While I started with `ds3231_` as the prefix, and that has prior art in tree for (e.g.) `ccs811_`, Zephyr has not carefully managed namespaces and I don't want to assume that there will never be conflicts between device model numbers across vendors.  Standardizing on the compatible name makes sense.  The negative impact is small, only affecting those who are using the custom API.

This proposed policy is something we should discuss more widely.  So: #24978

`callback` is unambiguous while `cb` is only clear in context with interpretation.  IMO the biggest problem with long-but-clear names is a consequence of 8-space indentation with a 78 character line limit.  I have no reason to believe the shorter version ever directly improves comprehension, and some reason to believe it makes it worse (increased cognitive load from requiring an interpretation at all).

I'm seeing 44 `foo_callback_t` typedefs, and 96 `foo_cb_t` typedefs, in Zephyr today.

Not to say I won't change it if people insist, but that's why I picked the names that I did.

**Review by henrikbrixandersen — APPROVED** [review]
I have mainly reviewed the DS3231 specific API extensions. Looks good!

I think we should get this merged. Once we have more RTC drivers in-tree, we can look at extending the counter API in a generic fashion or even defining a new, RTC specific API.

File: samples/drivers/counter/maxim_ds3231/sample.yaml
Comment (pabigot) [reply samples/drivers/counter/maxim_ds3231/sample.yaml:6]: The ChronoDot isn't physically compatible with Arduino headers, and at least two boards that are supported don't have them either.  So I'm reluctant to do that.

File: doc/reference/peripherals/index.rst
Comment (pabigot) [reply doc/reference/peripherals/index.rst:30]: The documentation for the driver-specific extensions ends up here.  It does state that it's a placeholder.

File: samples/drivers/counter/maxim_ds3231/sample.yaml
Comment (sslupsky) [reply samples/drivers/counter/maxim_ds3231/sample.yaml:6]: > I think we should get this merged. Once we have more RTC drivers in-tree, we can look at extending the counter API in a generic fashion or even defining a new, RTC specific API.

There are a few drivers in the tree already.  Many of the existing SOC's have RTC devices on-chip.  Zephyr is using them for the system timer (as opposed to a counter).  So, in this regard, I think it is quite important to advance the discussion of an RTC api before it becomes a terrible mess to fix in the future.

**Comment by carlescufi** [comment]
@nordic-krch would you mind approving this if you have no objections?

**Comment by carlescufi** [comment]
@jfischer-phytec-iot @sslupsky @nordic-krch please approve if you are fine with this PR as-is.

**Comment by carlescufi** [comment]
Merging this since it's been open for far too long. If you have additional comments please voice them to @pabigot and they can go in as bug fixes during the 2.3 stabilization period.

**Comment by cederom** [comment]
This is a nice starting point thank you but also requires external chip. I have created a feature request here: https://github.com/zephyrproject-rtos/zephyr/issues/35333 :-)

**Pull request merged**
